### PR TITLE
re-introduced parse module

### PR DESCRIPTION
**Pull request by brennanjl**
As discussed this morning, this PR re-introduces the parse module. It actually moves quite a bit of the engine to the parse module (such as most of the procedural code and sqlanalyzer). I've made it a separate PR to procedures so that it is easy to tell what is new vs https://github.com/kwilteam/kwil-db/pull/608, since #608 has already been reviewed and for the most part approved.

Keep in mind that there are also some half-baked additions for foreign calls that I included here from a separate branch. Making this change would've totally screwed the foreign call branch, so I decided to include it here.

`parse` now features a top-level `ParseKuneiform` function, which will parse a kuneiform string, as well as all actions and procedures. It also performs cartesian join checks, type checking, etc..

Ideally, we would only let consumers import the top-level package in `parse`, however I am not sure if this is possible. I tried adding an `internal` package, but the main Kwil module was unable to import from there, so I had to leave it as-is.

### CODE REVIEW

**Comment by KwilLuke** [comment]
FYI @brennanjl, I think we need to add `foreign_calls` to proto before I can test this with kwil-js. I think it makes sense for me to wait for this and #669 to go to main before implementing the changes needed to support foreign_calls in kwil-js (assuming this will get merged at least a week before release).

File: internal/engine/ddl/foreign_procedure.go
Comment (Yaiba) [inline internal/engine/ddl/foreign_procedure.go:144]: I think this logic could use a bit refactor? Not sure if I get this logic right

```go
// if proc.Returns == nil
// ...
// logic for !Returns.IsTtable
// ...
// if proc.Returns.IsTable {
//   if len(proc.Returns.Fields) == 0{
//      str.WriteString(" RETURNS VOID")
//   }
//   str.WriteString(" RETURNS TABLE(")
//   ...
// }


```

**Review by Yaiba — CHANGES_REQUESTED** [review]
In general looks to me, there are some new json tags which are not using snake case and we shoud enforce it

File: core/types/schema.go
Comment (brennanjl) [reply core/types/schema.go:845]: Good point. I will make that change. @KwilLuke this will be a breaking change for you

File: core/types/schema.go
Comment (brennanjl) [reply core/types/schema.go:788]: @KwilLuke another breaking change for you

File: parse/go.mod
Comment (brennanjl) [reply parse/go.mod:9]: ! good catch

File: parse/parse.go
Comment (brennanjl) [reply parse/parse.go:17]: We actually do test this for procedures (we can't for actions).

Comment (brennanjl) [reply parse/parse.go:17]: See `parse/metadata/functions.go`. Function definitions have a `ValidateArgs` func.

Comment (Yaiba) [reply parse/parse.go:17]: Those are for functions right?
I mean to validate table column attributes, before it's send to server.  The example I mentioned is for something like 

```
database testdb;
table user {
 id int primary,
 name text min(3);
}
```

The table def is wrong

File: scratch
Comment (brennanjl) [reply scratch:1]: Oddly, I do not have this file locally

File: parse/parse.go
Comment (brennanjl) [reply parse/parse.go:17]: Oh oh oh I see.

Will add that, good catch

File: core/types/schema.go
Comment (Yaiba) [inline core/types/schema.go:800]: There are some tags, for example `isArrary`, should change those as well

File: core/types/schema.go
Comment (Yaiba) [inline core/types/schema.go:293]: This is good. 
Maybe also add index and foreign_key validation? i.e. make sure referred column exists. 
That's so far the validations I can think of that we can do for table definition.

**Comment by Yaiba** [comment]
This pr looks good to me now.  There is one last thing,  we might want to get rid of the `replace` in go mod before we merge?

@jchappelow are you reviewing this?

**Review by Yaiba — APPROVED** [review]

**Comment by jchappelow** [comment]
> This pr looks good to me now. There is one last thing, we might want to get rid of the `replace` in go mod before we merge?
> 
> @jchappelow are you reviewing this?

Yeah I'm switching to it now. Will just give it an hour.  O gather this is holding up other things

**Comment by jchappelow** [comment]
> This pr looks good to me now. There is one last thing, we might want to get rid of the `replace` in go mod before we merge?

I don't see how we can get rid of `replace`s now. require on `github.com/kwilteam/kwil-db/core v0.1.2` from parse.go isn't adequate.  eventually we'll switch that to require `core/v0.2.0` but not now

File: parse/go.mod
Comment (jchappelow) [inline parse/go.mod:9]: This almost defeats the whole purpose of making a parse module.  `parse` consumers don't want to require the main module, and we don't want a circular `parse` <-> main module require.

Comment (jchappelow) [reply parse/go.mod:9]: I see it used in parse/sql/sqlanalyzer/clean/clean.go for `validation.IsKeyword`.

I guess you wanted to use `core/types/validation.IsKeyword` instead?

I also see the only other users of `validation.IsKeyword` are `validation.ValidateIdentifier` and `validation.ValidateDBID`, and both of those seem to be unused?  (there's a also now a `core/types/validation.ValidateIdentifier` that seems to provide this now, so I bet the common/validation is now obsolete?

Comment (jchappelow) [reply parse/go.mod:9]: Ah, yeah, that's all it was.

`import "github.com/kwilteam/kwil-db/core/types/validation"

in parse/sql/sqlanalyzer/clean/clean.go

seems to remove this require.

File: parse/kuneiform/gen/kuneiform_lexer.go
Comment (jchappelow) [inline parse/kuneiform/gen/kuneiform_lexer.go:1]: Was a git rm from kuneiform/gen/kuneiform_lexer.go?

Comment (jchappelow) [reply parse/kuneiform/gen/kuneiform_lexer.go:1]: Oh, the diff must have exceeded it's threshold for matching.  All good.

File: parse/go.mod
Comment (brennanjl) [reply parse/go.mod:9]: Oh wow total mistake on my end. That old validation package shouldn't be there, definitely meant to use core

**Comment by Yaiba** [comment]
> > This pr looks good to me now. There is one last thing, we might want to get rid of the `replace` in go mod before we merge?
> 
> I don't see how we can get rid of `replace`s now. require on `github.com/kwilteam/kwil-db/core v0.1.2` from parse.go isn't adequate. eventually we'll switch that to require `core/v0.2.0` but not now

right, forgot we dont' have a tag on main works

File: parse/go.mod
Comment (brennanjl) [reply parse/go.mod:9]: TIL you can have circular dependencies between submodules

File: parse/kuneiform/gen/kuneiform_lexer.go
Comment (brennanjl) [reply parse/kuneiform/gen/kuneiform_lexer.go:1]: Ah I was not actually sure what was going on there

File: internal/engine/ddl/foreign_procedure.go
Comment (jchappelow) [inline internal/engine/ddl/foreign_procedure.go:106]: I think this function is not a hot path, called only on deploy, but for simple string concatenation it's often a good idea to go with `+`.  The `fmt` printers are relatively heavy duty, using lots of reflection and generally useless operations for simple tasks like this.

```go
str.WriteString(`CREATE OR REPLACE FUNCTION _fp_` + proc.Name + `(_dbid TEXT, _procedure TEXT`)
```

https://dev.to/pmalhaire/concatenate-strings-in-golang-a-quick-benchmark-4ahh

Again, probably not a worry here, but something to be aware of in code that is called a lot.

File: internal/engine/ddl/foreign_procedure.go
Comment (jchappelow) [inline internal/engine/ddl/foreign_procedure.go:134]: Maybe I'm reading this wrong because it's Friday afternoon, but this looks like it's going to write `_out__out_X`.

File: internal/engine/ddl/foreign_procedure.go
Comment (brennanjl) [reply internal/engine/ddl/foreign_procedure.go:106]: Ah thats good to know. None of the string formatting are hot paths as you mentioned, but I will note that for the future.

I probably won't bother to change this right now, since I actually think we will want to re-evaluate how procedures work under the hood after this release. translating to plpgsql is very hard, and extremely error prone. It will be near impossible to add gas to; I think it will require making it a postgres extension after this initial version.

File: internal/engine/ddl/foreign_procedure.go
Comment (brennanjl) [reply internal/engine/ddl/foreign_procedure.go:134]: Lol yes this is a bug

File: parse/sql/sqlanalyzer/typing/visitor.go
Comment (jchappelow) [inline parse/sql/sqlanalyzer/typing/visitor.go:86]: This unexported function appears to be unused.

File: parse/wasm/wasm.go
Comment (jchappelow) [inline parse/wasm/wasm.go:11]: Up a few lines, on the `//go:generate` line, there appears to be a typo, lacking a space between `netgo` and `wasm.go` (it's passing the tag `netgowasm.go` and the source is implied).

```diff
diff --git a/parse/wasm/wasm.go b/parse/wasm/wasm.go
index d3ba98ce..19413e1e 100644
--- a/parse/wasm/wasm.go
+++ b/parse/wasm/wasm.go
@@ -1,6 +1,6 @@
 //go:build js && wasm
 
-//go:generate env GOOS=js GOARCH=wasm go build -o kuneiform.wasm -ldflags "-s -w" -trimpath -tags netgowasm.go
+//go:generate env GOOS=js GOARCH=wasm go build -o kuneiform.wasm -ldflags "-s -w" -trimpath -tags netgo wasm.go
 package main
```

Probably good to add ` CGO_ENABLED=0` too.

File: internal/engine/ddl/foreign_procedure_test.go
Comment (jchappelow) [inline internal/engine/ddl/foreign_procedure_test.go:95]: Need to keep this one?

File: internal/engine/execution/procedure.go
Comment (jchappelow) [inline internal/engine/execution/procedure.go:161]: Any way this can fail to fall the `calledAction`, leaving it nil?

File: internal/engine/execution/procedure.go
Comment (jchappelow) [inline internal/engine/execution/procedure.go:399]: `makeExecutables` doesn't have a possible error return, although the `evaluatables` do

File: internal/engine/execution/procedure.go
Comment (jchappelow) [inline internal/engine/execution/procedure.go:406]: Is this append for a one line slice copy?  Maybe `slices.Clone` would be more straightforwared, but I think a deep copy of the fields is more than what's needed to avoid the loop variable capture.

File: internal/engine/execution/queries.go
Comment (jchappelow) [inline internal/engine/execution/queries.go:101]: Oh yeah, I thought this was in the last PR.

File: internal/engine/integration/schema_test.go
Comment (jchappelow) [inline internal/engine/integration/schema_test.go:41]: uncomment or delete?

File: parse/kuneiform/grammar/KuneiformLexer.g4
Comment (jchappelow) [inline parse/kuneiform/grammar/KuneiformLexer.g4:106]: In `.*?`, what is the `?` for?  At least in regexp, `*` means zero or more.

**Review by jchappelow — CHANGES_REQUESTED** [review]
Looks pretty good.  Need a few updates to CI and tooling though:

In .golangci.yml:


```diff
diff --git a/.golangci.yml b/.golangci.yml
index 42d61f4b..4c2ddafd 100644
--- a/.golangci.yml
+++ b/.golangci.yml
@@ -3,6 +3,10 @@ run:
   skip-dirs:
     - core/rpc/protobuf
     - core/rpc/http
+    - parse/sql/gen
+    - parse/procedures/gen
+    - parse/kuneiform/gen
+    - parse/actions/gen
```

Taskfile.yml lint and fmt updates:


```diff
diff --git a/Taskfile.yml b/Taskfile.yml
index 480b4f71..b38ed18f 100644
--- a/Taskfile.yml
+++ b/Taskfile.yml
@@ -47,8 +47,12 @@ tasks:
       # so we have to list all the folders that are adjacent to protobuf.
       - |
         goimports -format-only -w ./cmd ./internal ./test ./core/client \
-          ./core/crypto ./core/log ./core/rpc/client ./core/rpc/conversion \
-          ./core/rpc/transport ./core/types ./core/utils
+          ./core/crypto ./core/log ./core/rpc/client \
+          ./core/rpc/transport ./core/types ./core/utils \
+          ./parse/parse.go ./parse/metadata ./parse/kuneiform/*.go \
+          ./parse/procedures/*.go ./parse/procedures/parser ./parse/procedures/visitors \
+          ./parse/sql/*.go ./parse/sql/postgres ./parse/sql/sqlanalyzer ./parse/sql/tree \
+          ./parse/actions/*.go ./parse/actions/parser \
 
   tidy:
     desc: go mod tidy each module
@@ -64,7 +68,7 @@ tasks:
     desc: Lint with golangci-lint
     cmds:
       # skip-dirs only works in .yaml
-      - golangci-lint run ./... ./core/... ./test/... -c .golangci.yml
+      - golangci-lint run ./... ./core/... ./test/... ./parse/... -c .golangci.yml
```

Also, `git rm internal/parse/sql/tree/functions.md`?

File: internal/engine/execution/procedure.go
Comment (brennanjl) [reply internal/engine/execution/procedure.go:399]: I don't think so, but I will add a check just in case

File: internal/engine/execution/procedure.go
Comment (brennanjl) [reply internal/engine/execution/procedure.go:161]: I don't think so, but I will add a check just in case

File: internal/engine/execution/procedure.go
Comment (brennanjl) [reply internal/engine/execution/procedure.go:406]: Did not know about `slices.Clone`.

However you are right, I do not think we actually have to copy the `OrderedParams` slice.

File: internal/engine/execution/queries.go
Comment (brennanjl) [reply internal/engine/execution/queries.go:101]: Could've sworn it was. A lot of this has sort've been continual work, as the first few days of foreign calls were being worked on while procedures was in review.

Probably just got my wires crossed here

File: parse/kuneiform/grammar/KuneiformLexer.g4
Comment (brennanjl) [reply parse/kuneiform/grammar/KuneiformLexer.g4:106]: This seems to be the way of doing block comments in antlr. I noticed Gavin had used it before, chatgpt used it, others use it: https://stackoverflow.com/questions/12898052/antlr-how-to-skip-multiline-comments

File: parse/wasm/wasm.go
Comment (jchappelow) [inline parse/wasm/wasm.go:3]: `CGO_ENABLED=0` would go with the other `env` settings like `GOOS` rather.

Comment (brennanjl) [reply parse/wasm/wasm.go:3]: whoops thanks. Sorry, been a long day

**Review by jchappelow — APPROVED** [review]

**Pull request merged**
